WordPress/gutenberg
# [WIP] Patterns: Explore partial syncing

Related Issues:
- https://github.com/WordPress/gutenberg/discussions/50456
- https://github.com/WordPress/gutenberg/issues/53705

Related explorations:
- https://github.com/WordPress/gutenberg/pull/53789
- https://github.com/WordPress/gutenberg/pull/51375
- https://github.com/WordPress/gutenberg/pull/53247

## What?

Explores possibilities for partially syncing patterns. 

**At present, this is more Frankenstein than a proper solution**. It hacks together a few different approaches. 

Some key aspects to note are:

- The needs for partial syncing seem a bit off compared to block connections as rather than replacing or redefining an attribute, we need to define and alternative, preferred, source.
- Still requires an update to block.json files to opt into pattern support (and define attributes that can be customized per pattern instance)
- Stores a unique id per block in a source pattern so that customized attributes or content on instances can be assigned to the correct block within the pattern
- Leverages context similar to https://github.com/WordPress/gutenberg/pull/53789 to pass customizations from the pattern instance down to inner blocks to override their defaults
- Sets up the pattern support similar to block supports and uses a filter to update block attributes to opt into receiving the dynamic content context from a pattern instance
- Uses a `render_block` filter to process the customized content/attributes and update a block's final markup

## Important

**This is just a rough prototype hard coded to gain understanding - it is not intended for production implementation**

- Only works for paragraph content and image URLs at this point
- There are already several known issues (e.g. blocks within patterns need locking). These will be iterated on should this exploration form the basis of an implementable approach


## Why?

See https://github.com/WordPress/gutenberg/issues/53705 for reasoning

## How?

This exploration has been added as a Gutenberg experiment which needs to be opted into to test.
 
<img width="400" alt="Screenshot 2023-08-24 at 5 09 49 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/29bc5b78-723d-4405-bb72-8b196a189551">

TBA when the approach stabilizes.

## Testing Instructions

<details>
<summary>1. Basic Use Case</summary>

- Open Appearance > Editor > Patterns > My patterns
- Create a simple synced pattern with only a paragraph within it
- Add some text to the pattern’s paragraph and save the pattern
- Load the post editor, create a new post and insert your new synced pattern
- Confirm that the pattern loads the correct default text, the block markup in the code editor view etc
- Save the post and view it on the frontend, confirming the paragraphs text is still correct
- Back in the post editor, select the inner paragraph block, and update the text
- Save the updates
- Reloading the post editor and the frontend, the update text displays correctly, no errors
- Switch to the code editor view and check that the dynamic content is saved under the same metadata.id as is in the source pattern’s paragraph.
- Duplicate the pattern, there should be no errors and inspecting the block markup in the code editor should show the same dynamic content attribute on the pattern
- Update the text in either pattern instance
- Save the post. This time there should be no prompt to save the pattern entity as the paragraph block in the source pattern already has the `metadata.id` attribute
- Confirm the correct text displays on the frontend as well
</details>

<details>
<summary>2. Updating source pattern content</summary>

- Appearance > Editor > Patterns > My patterns
- Create a new synced pattern, nesting a paragraph block inside a group block
- Give the group block a background color and some padding
- Save the pattern
- Open the post editor and insert the new pattern
- Update the inner paragraph’s text
- Confirm it displays correctly on the frontend as well
- Duplicate the pattern instance
- Save the post
- Switch back to the pattern editor and update the block’s background color
- Switch to the post editor and frontends and confirm the pattern instances there reflect the source pattern change
- Delete the custom paragraph text in one pattern, it should revert to the default in the source pattern
    - KNOWN ISSUE: Can’t fully clear the default content. Should that be possible? Probably, it makes it awkward deleting all the example content to then edit it further…
</details>

<details>
<summary>3.  Multiple pattern-supported inner blocks</summary>

- Appearance > Editor > Patterns > My patterns
- Create a pattern, with multiple blocks within it. Include an image and multiple paragraph blocks
- Make sure that the different paragraph blocks can be updated independently and display correctly on the frontend
- KNOWN ISSUE: If you delete a block, it changes the source pattern
- KNOWN ISSUE: If you duplicate a block in the pattern editor and a pattern instance has customizations for the original block, they are applied to the duplicated block as well because it duplicated the `metadata.id` attribute
</details>

<details>
<summary>4. Image url attribute</summary>

- Appearance > Editor > Patterns > My patterns
- Create a pattern with an image block within it
- Assign an image to the image block and save the pattern
- Add the pattern to a post via the post editor
- Save the post and confirm the correct image is there on the frontend
- Back in the post editor, replace the image 
- Save the post and confirm it has updated on the frontend
- Switch to the pattern editor and confirm the original image is present in the source pattern

</details>

Pattern support for heading block has also been added. Test overriding heading content on a per-pattern-instance basis using different levels of headings e.g. `h1-h6`.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/60436221/bbc828f2-9ec2-4364-bcfc-0de74053866b
